File: frontend/src/lib/ipc/remotes.ts
import { Channel } from "@tauri-apps/api/core";
import { commands, type RemoteInfo, type RepoId } from "./bindings";
import { unwrap } from "./index";

export type { RemoteInfo } from "./bindings";

/** The menu of a remote in Branches (doc/04-ipc-contract.md, «Remotes»). */
export async function remoteInfo(repo: RepoId, name: string): Promise<RemoteInfo> {
  return unwrap(await commands.remoteInfo(repo, name));
}

export async function renameRemote(repo: RepoId, from: string, to: string) {
  return unwrap(await commands.renameRemote(repo, from, to));
}

export async function removeRemote(repo: RepoId, name: string) {
  return unwrap(await commands.removeRemote(repo, name));
}

export async function setRemoteProperties(repo: RepoId, name: string, url: string, backgroundFetch: boolean) {
  return unwrap(await commands.setRemoteProperties(repo, name, url, backgroundFetch));
}

function lines(onLine: (line: string) => void): Channel<string> {
  const channel = new Channel<string>();
  channel.onmessage = onLine;
  return channel;
}

/** `false`: nothing new came. */
export async function fetchMore(repo: RepoId, remote: string, onLine: (line: string) => void) {
  return unwrap(await commands.fetchMore(repo, remote, lines(onLine)));
}

export async function fetchDepth(repo: RepoId, remote: string, depth: number, onLine: (line: string) => void) {
  return unwrap(await commands.fetchDepth(repo, remote, depth, lines(onLine)));
}

File: src-tauri/src/commands/remotes.rs
//! A remote's own menu in Branches: Properties, Rename, Delete, Fetch More, Set Depth (M5).

use super::{blocking, mutating_titled};
use app_state::{OperationKind, RepoId};
use git_engine::{GitError, RemoteInfo};

#[tauri::command]
#[specta::specta]
pub async fn remote_info(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    name: String,
) -> Result<RemoteInfo, GitError> {
    let app_state = state.state.clone();
    blocking("remote_info", move || app_state.remote_info(repo, &name)).await
}

#[tauri::command]
#[specta::specta]
pub async fn rename_remote(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    from: String,
    to: String,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Other,
        "Renaming the remote",
        "rename_remote",
        move || app_state.rename_remote(repo, &from, &to),
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn remove_remote(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    name: String,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Other,
        "Deleting the remote",
        "remove_remote",
        move || app_state.remove_remote(repo, &name),
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn set_remote_properties(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    name: String,
    url: String,
    background_fetch: bool,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Other,
        "Saving the remote",
        "set_remote_properties",
        move || app_state.set_remote_properties(repo, &name, &url, background_fetch),
    )
    .await
}

/// `false`: nothing new came from the remote.
#[tauri::command]
#[specta::specta]
pub async fn fetch_more(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    remote: String,
    on_progress: tauri::ipc::Channel<String>,
) -> Result<bool, GitError> {
    let app_state = state.state.clone();
    super::network::networking(
        &state.state,
        repo,
        OperationKind::Fetch,
        "fetch_more",
        move |stop| {
            super::network::with_progress("fetch", &remote, &on_progress, |on_line| {
                app_state.fetch_more(repo, &remote, &stop, on_line)
            })
        },
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn fetch_depth(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    remote: String,
    depth: u32,
    on_progress: tauri::ipc::Channel<String>,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    super::network::networking(
        &state.state,
        repo,
        OperationKind::Fetch,
        "fetch_depth",
        move |stop| {
            super::network::with_progress("fetch", &remote, &on_progress, |on_line| {
                app_state.fetch_depth(repo, &remote, depth, &stop, on_line)
            })
        },
    )
    .await
}
